cronjob_scripts/crates_io_popular_crates.py

```diff
@@ -0,0 +1,97 @@
+    def __init__(self, tarball: tarfile.TarFile, member: tarfile.TarInfo):
+        self.tarball = tarball
+        self.member = member
+
+    def is_one_of_csvs_interested(self, csvs_to_extract: tuple[str, ...]) -> str | None:
+        if not self.member.isfile():
+            return None
+        for name in csvs_to_extract:
+            if self.member.name.endswith(f"data/{name}"):
+                return name
+        return None
+
+    def get_file_stream(self):
+        stream = self.tarball.extractfile(self.member)
+        assert stream
+        return stream
+
+def download_tar_gz(url):
+    response = requests.get(url, stream=True)
+    response.raise_for_status()
+
+    with tarfile.open(fileobj=response.raw, mode='r:gz') as tarball:
+        while True:
+            member = tarball.next()
+            if member is None:
+                return
+            yield TarEntry(tarball, member)
+
+def get_crates_io_popular_crates_inner(minimum_downloads=200000):
+    files_to_extract = ("crate_downloads.csv", "crates.csv", "default_versions.csv", "versions.csv")
+    files_extracted = 0
+    dfs = {}
+    for entry in download_tar_gz('https://static.crates.io/db-dump.tar.gz'):
+        name = entry.is_one_of_csvs_interested(files_to_extract)
+        if name:
+            dfs[name] = pl.scan_csv(entry.get_file_stream())
+            files_extracted += 1
+            if files_extracted == len(files_to_extract):
+                return (
+                    dfs["crate_downloads.csv"]
+                    .join(dfs["crates.csv"].select("id", "name"), left_on="crate_id", right_on="id")
+                    .join(dfs["default_versions.csv"], on="crate_id")
+                    .join(
+                        dfs["versions.csv"].select("id", "crate_id", "yanked", "bin_names"),
+                        left_on=("crate_id", "version_id"),
+                        right_on=("crate_id", "id"),
+                    )
+                    .filter(pl.col("bin_names") != "{}", pl.col("yanked") == "f")
+                    .filter(pl.col("downloads") > minimum_downloads)
+                    .select("name")
+                )
+
+    raise RuntimeError(f"Failed to find all csvs {files_to_extract}")
+
+
+def get_crates_io_popular_crates(minimum_downloads=200000):
+    cached_path = Path("cached_crates_io_popular_crates.parquet")
+    if cached_path.is_file():
+        # TODO: Once `iter_rows()` can be used on LazyFrame, use it for better perf
+        #  and less ram usage.
+        # https://github.com/pola-rs/polars/issues/10683
+        df = pl.scan_parquet(cached_path).collect(streaming=True)
+    else:
+        df = (
+            get_crates_io_popular_crates_inner(minimum_downloads)
+            # TODO: Use streaming, maybe use sink_parquet instead?
+            # https://github.com/pola-rs/polars/issues/18684
+            .collect()
+        )
+        df.write_parquet(cached_path, statistics=False)
+    return (row[0] for row in df.iter_rows())
```

Comment:
I was wondering if we could share anything like my decorator at https://github.com/cargo-bins/cargo-quickinstall/pull/288/files#diff-5cb8ada8e1547a375244dba018abfe68cdee554ee84f24b6ee342b050878c3a0R105-R128



It feels like you're aggressively optimising for speed here. Remember that we have a sleep(30) in the inner loop when we are triggering packages, so speed probably doesn't matter **that** much.

I suspect you'd do better if you return the dataframe, and let the caller do `set(df["name"])`  or something(polars is a based columnar datastore, so getting the column as a series will probably be quicker). It might be nicer for the downstream code to work in terms of the polars.Series rather than converting to an iterator and collecting into a set(). I'm trying to avoid getting caught up micro-optimising this, because I think simplicity is best for the cronjob code.

Also remember that we will be calling this function once per architecture, so maybe it's best to just keep the dataframe in memory (I will probably be fixing that soon though, so maybe don't bother)

Reply:
... the parquet file is 2785B. We should definitely just be calling read_parquet() and returning the df rather than even thinking about being clever about it.

Reply:
I return an iterable here since the the previous `get_popular_crates()` return a `list`.

Returning a `DataFrame` is ok, but it doesn't make much sense for the caller, it just expects something that is iterable and can be passed to `set`

```python3
>>> pl.DataFrame({"name": ["a"]})
shape: (1, 1)
┌──────┐
│ name │
│ ---  │
│ str  │
╞══════╡
│ a    │
└──────┘
>>> df = pl.DataFrame({"name": ["a"]})
>>> list(row for row in df)
[shape: (1,)
Series: 'name' [str]
[
	"a"
]]
```

`pl.DataFrame` is not iterable`, `pl.DataFrame.iter_rows` returns `list[tuple[str]]`:

```python3
>>> pl.DataFrame({"name": ["a"]})
shape: (1, 1)
┌──────┐
│ name │
│ ---  │
│ str  │
╞══════╡
│ a    │
└──────┘
>>> df = pl.DataFrame({"name": ["a"]})
>>> list(row for row in df)
[shape: (1,)
Series: 'name' [str]
[
	"a"
]]
```

so it's better to return a generator of `Iterable[str]`.